lib/gc/Analysis/MatmulConfigAnalysis.cpp

```diff
@@ -0,0 +1,438 @@
+
+#include "gc/Analysis/MatmulConfigAnalysis.h"
+#include <limits>
+#include <llvm/Support/Debug.h>
+
+namespace mlir {
+namespace gc {
+
+#define DEBUG_TYPE "matmul-config-analysis"
+
+llvm::raw_ostream &operator<<(llvm::raw_ostream &ss,
+                              const MatmulConfig &config) {
+
+  ss << "MBlock: " << config.MBlock << ", NBlock: " << config.NBlock
+     << ", KBlock: " << config.KBlock << ", MThreads: " << config.MThreads
+     << ", NThreads: " << config.NThreads << ", KThreads: " << config.KThreads
+     << ", innerMostMBlock: " << config.innerMostMBlock
+     << ", innerMostNBlock: " << config.innerMostNBlock
+     << ", innerMostKBlock: " << config.innerMostKBlock;
+  return ss;
+}
+
+template <typename T>
+static llvm::raw_ostream &operator<<(llvm::raw_ostream &ss,
+                                     std::vector<T> arry) {
+  ss << "[";
+  for (auto [idx, a] : llvm::enumerate(arry)) {
+    if (idx != 0) {
+      ss << ", ";
+    }
+    ss << a;
+  }
+  ss << "]";
+  return ss;
+}
+
+// generate the candidate for the block size(factor of `num`, pow of 2 which is
+// less than `num`)
+std::vector<uint32_t>
+getCandidate(uint32_t num, uint32_t floor,
+             uint32_t ceil = std::numeric_limits<uint32_t>::max()) {
+  // factor
+  std::vector<uint32_t> candidates;
+  for (uint32_t i = 1; i <= num; i++) {
+    if (num % i == 0 && i <= ceil && i >= floor) {
+      candidates.push_back(i);
+    }
+  }
+  // the pow of 2
+  uint32_t candidate = 1U;
+  while (candidate < num && candidate <= ceil && candidate >= floor) {
+    candidates.push_back(candidate);
+    candidate *= 2;
+  }
+  // remove duplicate candidates
+  std::sort(candidates.begin(), candidates.end());
+  candidates.erase(std::unique(candidates.begin(), candidates.end()),
+                   candidates.end());
+  return candidates;
+}
+
+// check if the config is valid
+bool isValidConfig(const MatmulConfig &config, SystemDesc &sysDesc,
+                   ArrayRef<uint32_t> shape) {
+  if (config.innerMostMBlock == 0 || config.innerMostNBlock == 0 ||
+      config.innerMostKBlock == 0) {
+    return false;
+  }
+
+  if (shape[0] % config.innerMostMBlock != 0 ||
+      shape[1] % config.innerMostNBlock != 0 ||
+      shape[2] % config.innerMostKBlock != 0) {
+    return false;
+  }
+
+  return true;
+}
+
+// check if the threads are valid
+bool validateThreads(ArrayRef<uint32_t> threads, SystemDesc &sysDesc) {
+  uint32_t numThreads = sysDesc.getNumThreads();
+  uint32_t actualThreads = 1U;
+  for (uint32_t t : threads) {
+    actualThreads *= t;
+  }
+  return actualThreads == numThreads;
+}
+
+// calculate the cost of the hardware efficiency(whether the vector register is
+// fully utilized)
+double hardwareEfficiencyCost(linalg::LinalgOp &linalgOp,
+                              ArrayRef<uint32_t> shape,
+                              const MatmulConfig &config, SystemDesc &sysDesc) {
+  llvm::TypeSize dtypeSize = DataLayout().getTypeSizeInBits(
```

Comment:
Since it's only considering the vector register utilization, using `vectorRegEfficiencyCost` might be better?
Is it possible that the different in&out have different data type?

Reply:
The name has changed accordingly. The input and output should have the same data type now and the deepTileMatmul will break `bf16xbf16->bf16` to `bf16xbf16->f32+cast(f32->bf16)` internally.